dbcrossbarlib/src/drivers/bigquery_shared/column_name.rs

```diff
@@ -0,0 +1,206 @@
+//! BigQuery column names.
+
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+use std::{
+    cmp::Ordering,
+    convert::TryFrom,
+    fmt,
+    hash::{Hash, Hasher},
+    str::FromStr,
+};
+
+use crate::common::*;
+
+/// A BigQuery column name.
+///
+/// This behaves like a string that preserves case, but which ignores it for
+/// comparisons. It may only contain valid BigQuery column names.
+///
+/// According to the official docs:
+///
+/// > A column name must contain only letters (a-z, A-Z), numbers (0-9), or
+/// > underscores (_), and it must start with a letter or underscore. The
+/// > maximum column name length is 128 characters. A column name cannot use any
+/// > of the following prefixes:
+/// >
+/// > - _TABLE_
+/// > - _FILE_
+/// > - _PARTITION
+/// >
+/// > Duplicate column names are not allowed even if the case differs. For
+/// > example, a column named Column1 is considered identical to a column named
+/// > column1.
+///
+/// [docs]: https://cloud.google.com/bigquery/docs/schemas#column_names
+#[derive(Clone)]
+pub(crate) struct ColumnName {
+    /// The original, mixed-case string, followed by an all-lowercase copy.
+    ///
+    /// Since we know that ASCII strings always have one character per byte, and
+    /// that lowercasing a string doesn't change its length, we can assume that
+    /// the dividing point is always exactly in the middle.
+    data: String,
+}
+
+impl ColumnName {
+    /// The original string, including case information.
+    pub(crate) fn as_str(&self) -> &str {
+        &self.data[..self.data.len() / 2]
+    }
+
+    /// Am all-lowecase version
+    fn as_lowercase(&self) -> &str {
+        &self.data[self.data.len() / 2..]
```

Comment:
wow that does not look like "make me lowercase"

Reply:
Did you see the comment on the `data` member? We pre-compute the lowercase string and append it to the regular string, and then use either the first or second half of the buffer via `as_str` or `as_lowercase`.